lightning/src/util/ser_macros.rs

```diff
@@ -1,3 +1,91 @@
+macro_rules! encode_tlv {
+	($stream: expr, {$(($type: expr, $field: expr)),*}) => { {
+		use bitcoin::consensus::Encodable;
+		use bitcoin::consensus::encode::{Error, VarInt};
+		use util::ser::{WriterWriteAdaptor, LengthCalculatingWriter};
+		$(
+			VarInt($type).consensus_encode(WriterWriteAdaptor($stream))
+				.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+			let mut len_calc = LengthCalculatingWriter(0);
+			$field.write(&mut len_calc)?;
+			VarInt(len_calc.0 as u64).consensus_encode(WriterWriteAdaptor($stream))
+				.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+			$field.write($stream)?;
+		)*
+	} }
+}
+
+macro_rules! encode_varint_length_prefixed_tlv {
+	($stream: expr, {$(($type: expr, $field: expr)),*}) => { {
+		use bitcoin::consensus::Encodable;
+		use bitcoin::consensus::encode::{Error, VarInt};
+		use util::ser::{WriterWriteAdaptor, LengthCalculatingWriter};
+		let mut len = LengthCalculatingWriter(0);
+		encode_tlv!(&mut len, {
+			$(($type, $field)),*
+		});
+		VarInt(len.0 as u64).consensus_encode(WriterWriteAdaptor($stream))
+			.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+		encode_tlv!($stream, {
+			$(($type, $field)),*
+		});
+	} }
+}
+
+macro_rules! decode_tlv {
+	($stream: expr, {$(($reqtype: expr, $reqfield: ident)),*}, {$(($type: expr, $field: ident)),*}) => { {
+		use ln::msgs::DecodeError;
+		let mut max_type: u64 = 0;
+		'tlv_read: loop {
+			use bitcoin::consensus::encode;
+			use util::ser;
+			use std;
+
+			let typ: encode::VarInt = match encode::Decodable::consensus_decode($stream) {
+				Err(encode::Error::Io(ref ioe)) if ioe.kind() == std::io::ErrorKind::UnexpectedEof
+					=> break 'tlv_read,
+				Err(encode::Error::Io(ioe)) => Err(DecodeError::from(ioe))?,
+				Err(_) => Err(DecodeError::InvalidValue)?,
+				Ok(t) => t,
+			};
+			if typ.0 == std::u64::MAX || typ.0 + 1 <= max_type {
+				Err(DecodeError::InvalidValue)?
+			}
+			$(if max_type < $reqtype + 1 && typ.0 > $reqtype {
+				Err(DecodeError::InvalidValue)?
+			})*
+			max_type = typ.0 + 1;
```

Comment:
The `+ 1` throughout here is a bit confusing. Is it needed because otherwise you couldn't support type `0` with this logic? Isn't `std::u64::MAX` a valid type, though?

The name `max_type` can be confusing, too. While it is the maximum type seen, the logic is such that it's really the minimum type that can be read next. Naming as `min_next_type` may read better with the logic. So I think this could be simplified and would read better as:

```rust
let mut last_seen_type: Option<u64> = None;
loop {
    let min_next_type = match last_seen_type {
        Some(x) => x + 1,
        None => 0,
    };

    // Types must be monotonically increasing.
    if typ.0 < min_next_type {
        Err(DecodeError::InvalidValue)?
    }

    // Required types must be found.
    $(if min_next_type <= $reqtype && typ.0 > $reqtype {
        Err(DecodeError::InvalidValue)?
    })*

    last_seen_type = Some(typ.0);
}
```

Reply:
Right, indeed, I suppose u64::MAX is allowed since its odd :/. I think I was trying to be too clever...I took a variation on your proposal (because I think your proposal would panic with an addition overflow if last_seen_type == u64::MAX due to the +1 in the Some branch).